paprika/tleap.py

```diff
@@ -0,0 +1,574 @@
+            raise Exception("The `buffer_target` should either be a string ending with 'A' (e.g., 12A) for 12 Angstroms of "
+                            "buffer or an int (e.g., 2000) for 2000 waters.")
+
+    def count_waters(self):
+        """
+        Quickly check the number of waters added for a given buffer size.
+
+        Returns
+        -------
+        waters : int
+
+        """
+        waters = self.count_residues()['WAT']
+        return waters
+
+    def count_residues(self):
+        """
+        Run and parse `tleap` output and return a dictionary of residues in the structure.
+    
+        Returns
+        -------
+        residues : dict
+            Dictionary of added residues and their number
+
+        """
+        self.write_input(include_saves=False)
+        output = self.run()
+        # Return a dictionary of {'RES' : number of RES}
+        residues = {}
+        for line in output[0].splitlines():
+            # Is this line a residue from `desc` command?
+            match = re.search("^R<(.*) ", line)
+            if match:
+                residue_name = match.group(1)
+                # If this residue is not in the dictionary, initialize and
+                # set the count to 1.
+                if residue_name not in residues:
+                    residues[residue_name] = 1
+                # If this residue is in the dictionary, increment the count
+                # each time we find an instance.
+                elif residue_name in residues:
+                    residues[residue_name] += 1
+        #log.debug(residues)
+        return residues
+
+    def set_additional_ions(self):
+        """
+        Determine whether additional ions (instead of or in addition to neutralization) are requested...
+    
+        Sets
+        -------
+        self.add_ion_residues : list
+            A processed list of ions and their amounts that can be passed to `tleap`
+
+        """
+        if not self.add_ions:
+            return None
+        if len(self.add_ions) < 2:
+            raise Exception("No amount specified for additional ions.")
+        if len(self.add_ions) % 2 == 1:
+            raise Exception("The 'add_ions' list requires an even number of elements. "
+                            "Make sure there is a residue mask followed by a value for "
+                            "each ion to be added (or molarity ending in 'M' or molality ending in 'm').")
+        self.add_ion_residues = []
+        for ion, amount in zip(self.add_ions[0::2], self.add_ions[1::2]):
+            self.add_ion_residues.append(ion)
+            if isinstance(amount, int):
+                self.add_ion_residues.append(amount)
+            elif isinstance(amount, str) and amount[-1] == 'm':
+                # User specifies molality...
+                # number to add = (molality) x (number waters) x (0.018 kg/mol per water)
+                number_to_add = int(np.ceil(float(amount[:-1]) * self.target_waters * 0.018))
```

Comment:
This line assumes we're going to solvate with water (my fault). Maybe it should be configurable in case some daring person wants to specify ions in molality and solvent of CH3Cl.

Reply:
Good point.  Advanced variable.

Reply:
Done. Now `self.kg_per_mol_solvent`